pkg/ingester/ingester.go

```diff
@@ -218,13 +220,80 @@ func New(cfg Config, clientConfig client.Config, store ChunkStore, limits *valid
 	i.limiter = NewLimiter(limits, i.lifecycler, cfg.LifecyclerConfig.RingConfig.ReplicationFactor)
 
 	i.Service = services.NewBasicService(i.starting, i.running, i.stopping)
+
+	i.setupAutoForget()
+
 	return i, nil
 }
 
 func (i *Ingester) SetChunkFilterer(chunkFilter storage.RequestChunkFilterer) {
 	i.chunkFilter = chunkFilter
 }
 
+// setupAutoForget looks for ring status if `AutoForgetUnhealthy` is enabled
+// when enabled, unhealthy ingesters that reach `ring.kvstore.heartbeat_timeout` are removed from the ring every `HeartbeatPeriod`
+func (i *Ingester) setupAutoForget() {
+	if !i.cfg.AutoForgetUnhealthy {
+		return
+	}
+
+	go func() {
+		ctx := context.Background()
+		err := i.Service.AwaitRunning(ctx)
+		if err != nil {
+			level.Error(util_log.Logger).Log("msg", fmt.Sprintf("autoforget received error %s, autoforget is disabled", err.Error()))
+			return
+		}
+
+		level.Info(util_log.Logger).Log("msg", fmt.Sprintf("autoforget is enabled and will remove unhealthy instances from the ring after %v with no heartbeat", i.cfg.LifecyclerConfig.RingConfig.HeartbeatTimeout))
+
+		ticker := time.NewTicker(i.cfg.LifecyclerConfig.HeartbeatPeriod)
+		defer ticker.Stop()
+
+		for range ticker.C {
+			err := i.lifecycler.KVStore.CAS(ctx, ring.IngesterRingKey, func(in interface{}) (out interface{}, retry bool, err error) {
+				if in == nil {
+					return nil, false, nil
+				}
+
+				ringDesc, ok := in.(*ring.Desc)
+				if !ok {
+					level.Warn(util_log.Logger).Log("msg", fmt.Sprintf("ingester autoforget saw a KV store value that was not `ring.Desc`, got `%T`", in))
+					return nil, false, nil
+				}
+				var removeList []string
+				for id, ingester := range ringDesc.Ingesters {
+					if !ingester.IsHealthy(ring.Reporting, i.cfg.LifecyclerConfig.RingConfig.HeartbeatTimeout, time.Now()) {
+						if i.lifecycler.ID == id {
+							level.Warn(util_log.Logger).Log("msg", fmt.Sprintf("autoforget has seen our ID `%s` as unhealthy in the ring, network may be partitioned, skip forgeting ingesters this round", id))
+							return nil, false, nil
+						}
+						removeList = append(removeList, id)
+					}
+				}
+
+				if len(removeList) == len(ringDesc.Ingesters)-1 {
+					level.Warn(util_log.Logger).Log("msg", fmt.Sprintf("autoforget have seen %d unhealthy ingesters out of %d, network may be partioned, skip forgeting ingesters this round", len(removeList), len(ringDesc.Ingesters)))
+					return nil, false, nil
+				}
+
+				if len(removeList) > 0 {
+					for _, id := range removeList {
+						level.Info(util_log.Logger).Log("msg", fmt.Sprintf("forgeting ingester %v because it was not healthy after %v", id, i.cfg.LifecyclerConfig.RingConfig.HeartbeatTimeout))
+						ringDesc.RemoveIngester(id)
```

Comment:
I think it would be great to increment a metric counter here so that we can alert if this is happening too often for a long period. WDYT ?

Reply:
Yes! I do agree this is required to follow what's going on. Will add this.